### PR TITLE
New api to support Two form in one page.

### PR DESCRIPTION
**Pull request by Anu1601CS**
### Summary of Changes
New API to support custom form with different id's. It supports more than two form in one page.
Mulitiselect  :  `JHtml::_('behavior.multiselect', 'your id ');`  
like take a id = `updateform`
 
`JHtml::_('behavior.multiselect', 'updateForm');`

Grid id : `HTMLHelper::_('grid.id', $i, $value->hash_id, 'updateForm');`

Publish : `JHtml::_('jgrid.published', $value->state, $i, 'template.', 1, 'updateForm');`

Tool Bar : `ToolbarHelper::custom('template.deleteHistory', 'delete', 'move', 'JTOOLBAR_DELETE', true, 'updateForm');`

Pass your form id as a parameter.

Hope this ok. But the review is necessary. 
### Testing Instructions
Add an admin form with id = `adminfrom` and add the second form with id=`what ever you want`. In one page 

### Expected result
All form working as expected.

### Actual result
Old API only select the first form in the page with id = `adminform`. Which causes the error in working on the second form.

### Documentation Changes Required
Yes

@dgrammatiko Please take a look here. Is any thing left to support custom form id ?

### CODE REVIEW

File: libraries/cms/html/grid.php [hunk 1]
```diff
@@ -142,13 +142,20 @@ public static function checkall($name = 'checkall-toggle', $tip = 'JGLOBAL_CHECK
 	 * @param   boolean  $checkedOut  True if item is checked out
 	 * @param   string   $name        The name of the form element
 	 * @param   string   $stub        The name of stub identifier
+	 * @param   string   $form        An optional form selector.
 	 *
 	 * @return  mixed    String of html with a checkbox if item is not checked out, null if checked out.
 	 *
 	 * @since   1.5
 	 */
-	public static function id($rowNum, $recId, $checkedOut = false, $name = 'cid', $stub = 'cb')
```
Comment (dgrammatiko) [inline libraries/cms/html/grid.php]: You need to add the new parameter at the end or you’re breaking b/c

File: libraries/cms/html/jgrid.php [hunk 2]
```diff
@@ -155,7 +168,7 @@ public static function state($states, $value, $i, $prefix = '', $enabled = true,
 	 * @see     JHtmlJGrid::state()
 	 * @since   1.6
 	 */
-	public static function published($value, $i, $prefix = '', $enabled = true, $checkbox = 'cb', $publish_up = null, $publish_down = null)
+	public static function published($value, $i, $prefix = '', $enabled = true, $formId = null, $checkbox = 'cb', $publish_up = null, $publish_down = null)
```
Comment (dgrammatiko) [inline libraries/cms/html/jgrid.php]: Same here

File: libraries/cms/html/jgrid.php [hunk 3]
```diff
@@ -242,14 +255,15 @@ public static function published($value, $i, $prefix = '', $enabled = true, $che
...
+	 * @param   string        $formId    An optional form selector.
 	 * @param   string        $checkbox  An optional prefix for checkboxes.
 	 *
 	 * @return  string  The HTML markup
 	 *
 	 * @see     JHtmlJGrid::state()
 	 * @since   1.6
 	 */
-	public static function isdefault($value, $i, $prefix = '', $enabled = true, $checkbox = 'cb')
+	public static function isdefault($value, $i, $prefix = '', $enabled = true, $formId = null, $checkbox = 'cb')
```
Comment (dgrammatiko) [inline libraries/cms/html/jgrid.php]: And here

File: libraries/cms/html/jgrid.php [hunk 4]
```diff
@@ -320,13 +334,14 @@ public static function publishedOptions($config = array())
...
 	 * @param   string|array  $prefix      An optional task prefix or an array of options
 	 * @param   boolean       $enabled     True to enable the action.
+	 * @param   string        $formId      An optional form selector.
 	 * @param   string        $checkbox    An optional prefix for checkboxes.
 	 *
 	 * @return  string  The HTML markup
 	 *
 	 * @since   1.6
 	 */
-	public static function checkedout($i, $editorName, $time, $prefix = '', $enabled = false, $checkbox = 'cb')
```
Comment (dgrammatiko) [inline libraries/cms/html/jgrid.php]: And here

File: libraries/cms/html/jgrid.php [hunk 5]
```diff
@@ -354,13 +370,14 @@ public static function checkedout($i, $editorName, $time, $prefix = '', $enabled
...
+	 * @param   string        $formId    An optional form selector.
 	 * @param   string        $checkbox  An optional prefix for checkboxes.
 	 *
 	 * @return  string  The HTML markup
 	 *
 	 * @since   1.6
 	 */
-	public static function orderUp($i, $task = 'orderup', $prefix = '', $text = 'JLIB_HTML_MOVE_UP', $enabled = true, $checkbox = 'cb')
+	public static function orderUp($i, $task = 'orderup', $prefix = '', $text = 'JLIB_HTML_MOVE_UP', $enabled = true, $formId = null, $checkbox = 'cb')
 	{
```
Comment (dgrammatiko) [inline libraries/cms/html/jgrid.php]: And here

File: libraries/cms/html/jgrid.php [hunk 6]
```diff
@@ -382,13 +399,14 @@ public static function orderUp($i, $task = 'orderup', $prefix = '', $text = 'JLI
...
+	 * @param   string        $formId    An optional form selector.
 	 * @param   string        $checkbox  An optional prefix for checkboxes.
 	 *
 	 * @return  string  The HTML markup
 	 *
 	 * @since   1.6
 	 */
-	public static function orderDown($i, $task = 'orderdown', $prefix = '', $text = 'JLIB_HTML_MOVE_DOWN', $enabled = true, $checkbox = 'cb')
+	public static function orderDown($i, $task = 'orderdown', $prefix = '', $text = 'JLIB_HTML_MOVE_DOWN', $enabled = true, $formId = null, $checkbox = 'cb')
 	{
```
Comment (dgrammatiko) [inline libraries/cms/html/jgrid.php]: And here

**Comment by Anu1601CS** [comment]
@dgrammatiko Done! :)

File: libraries/cms/html/jgrid.php (hunk 5 above)
Comment (dgrammatiko) [reply libraries/cms/html/jgrid.php]: nope, it needs to be the last one, otherwise, we have a b/c break!!!

**Comment by Anu1601CS** [comment]
@dgrammatiko I think i need to create this PR to main repo right ?

**Comment by dgrammatiko** [comment]
Yes this needs to be done against the 4.0-dev

File: libraries/src/Toolbar/Button/StandardButton.php [hunk 7]
```diff
@@ -62,24 +62,36 @@ protected function prepareOptions(array &$options)
...
 	 * @deprecated  5.0 Use render() instead.
 	 */
-	public function fetchButton($type = 'Standard', $name = '', $text = '', $task = '', $list = true)
+	public function fetchButton($type = 'Standard', $name = '', $text = '', $task = '', $list = true, $formId = null)
 	{
-		$this->name($name)
-			->text($text)
-			->task($task)
-			->listCheck($list);
+		if ($formId !== null)
```
Comment (astridx) [inline libraries/src/Toolbar/Button/StandardButton.php]: Can we save some lines if we do it like this
 ```
...
if ($formId !== null)
{
	$this->form($formId);
}
$this->name($name);
$this->text($text);
$this->task($task);
$this->listCheck($list);
...
```

**Comment by astridx** [comment]
I have tested this.
1. I changed id and name of the form in file `/administrator/components/com_installer/tmpl/discover/default.php` to `name="adminForm1" id="adminForm1"` -> The form did not work any more. I could not install a discovered extension.
2. I changed the line `JHtml::_('behavior.multiselect');` to `JHtml::_('behavior.multiselect', 'adminForm1');`. Later I saw, that this is not needed.
3. I changed the line `<?php //echo JHtml::_('grid.id', $i, $item->extension_id); ?>` to `<?php echo HTMLHelper::_('grid.id', $i, $item->extension_id, false, 'cid', 'cb', 'adminForm1'); ?>`
4. Now it was possible for me to install a discovered extension

**Comment by astridx** [comment]
Now I am not sure how to merge this branch. Should we merge here in the project, or belongs it to the Main project?
@laoneo @dgrammatiko What do you think? Actually, it would belong to the main project, I think. But we can only continue working if we have these changes. How long would it take to merge this in the main project?

**Comment by Anu1601CS** [comment]
@astridx Yes, these changes is done! for our project. So, i think we need to merge here ?
@laoneo

**Comment by dgrammatiko** [comment]
Please make a or in the 4.0 branch and then we can merge it there. The code will be shortly available here then

**Pull request closed without merging**